### DISCUSSION TITLE
Consider enabling whitespace trimming in the Jinja2 env

### DISCUSSION THREAD

**Discussion by kamilkrzyskow** [mkdocs/mkdocs#3309] [Ideas]
Hi 👋,
kind of a weird feature to ask for, I know, but I got the idea after looking at the default source code of our built documentation page. We're using the Material theme, and for our deployment we're using the `minify` plugin too. 

But many people don't do that, either because they don't think about this, or don't want to increase their build time with the plugin. So their pages waste network and processing time globally ❗ 
However, I [checked](https://stackoverflow.com/questions/35775207/remove-unnecessary-whitespace-from-jinja-rendered-template) and it's possible to enable simple trimming directly in the env.

Testing it with a hook:
```py
def on_env(
    env: jinja2.Environment, *, config: MkDocsConfig, files: Files
) -> Optional[jinja2.Environment]:
    env.trim_blocks = True
    env.lstrip_blocks = True
    return env
```

`mkdocs new` with the Material theme got reduced from 11,795 down to 10,539 characters, so roughly a 10% reduction.
Our [homepage](https://gothic-modding-community.github.io/gmc/), which uses different features of Material etc. got reduced from 83,975 down to 65,449 characters, so roughly a 20% reduction,

The issue lies of course in the Material theme's lack of whitespace management, but this seems like the less painful solution in general if MkDocs would enable this Jinja2 feature by default.

The build time didn't change for me with this setting enabled.

EDIT:
Or perhaps it would be possible to add it as an opt-in feature for the themes 🤔

**Comment by kamilkrzyskow** [comment]
@oprypin Sorry for the mention, I just noticed that you're currently active. Could you please state your opinion on this matter?

**Comment by oprypin** [comment]
Sure it can be an opt-in feature for themes. Otherwise I think it can break stuff.

The importance of this is not *that* big, (same for minification actually) because everything should be served gzipped - you should measure that.

**Reply by kamilkrzyskow** [reply]
Thanks for taking a look.

I didn't think of HTML being served gzipped 🤔 I always thought it's "reserved" to assets like scripts or stylesheets.
So you're likely right, that it doesn't matter when it comes to size, but I shall check on GitHub Pages later to be sure.

However, there is also the aesthetic side of it, because when I open the raw HTML it has a lot of white space, and it's hard to read and check what's going on without using some Developer Tools or Prettify tools, but perhaps this is my personal grievance.

**Comment by pawamoy** [comment]
I think it's safe if a theme enables the `trim_blocks` and `lstrip_blocks` options, because this would only affect the HTML provided by the theme and not the HTML converted from Markdown.

**Reply by pawamoy** [reply]
I tried the `on_env` hook, and checked a big page of mine, it reduced lines by 20.6% and total characters by 3.4%. So there mostly were blank lines, coming from the theme itself, particularly in between nav and toc items. Minification on the other hand reduced lines by 97.7% (unsurprisingly) and characters by 15.3%, for an additional 0.5s during build time using [mkdocs-minify-html](https://github.com/monosans/mkdocs-minify-html-plugin) (instead of 2.5s with [mkdocs-minify-plugin](https://github.com/byrnereese/mkdocs-minify-plugin)).
